src/backend/distributed/metadata/metadata_cache.c

```diff
@@ -1164,6 +1164,25 @@ BuildCachedShardList(DistTableCacheEntry *cacheEntry)
 		}
 
 		heap_close(distShardRelation, AccessShareLock);
+
+		ShardInterval *firstShardInterval = shardIntervalArray[0];
+		bool foundInCache = false;
+		ShardCacheEntry *shardEntry = hash_search(DistShardCacheHash,
+												  &firstShardInterval->shardId, HASH_FIND,
+												  &foundInCache);
+		if (foundInCache)
+		{
+			/*
+			 * Normally, all shard cache entries for a given DistTableEntry are removed
+			 * before we get here. There is one exception: When a shard changes from
+			 * one relation ID to another. That typically happens during metadata
+			 * syncing when the distributed table is dropped and re-created without
+			 * changing the shard IDs. That means that old relation no longer exists
+			 * and we can safely wipe its entry, which will remove all corresponding
+			 * shard cache entries.
+			 */
+			ResetDistTableCacheEntry(shardEntry->tableEntry);
```

Comment:
Hmm, I couldn't find a way to trigger any bugs, but this seems a little bit dangerous to me.

We're in the middle of `BuildDistTableCacheEntry()`, and wouldn't calling `ResetDistTableCacheEntry ()` makes the `DistTableCacheEntry *` partially broken?

For example, `cacheEntry->partitionKeyString` is build before `BuildCachedShardList ()`, and now with this call we're invalidating `cacheEntry->partitionKeyString`.

Or, am I missing something?

Reply:
The tableEntry in the shardEntry is the table entry for the old table, while we're rebuilding the table entry for the new table. I can add an extra check to confirm this is the case. That way, we'll fall back to the old error if there's a bug, plus it makes the intent of the code slightly more clear.